packages/vscode-extension/src/project/project.ts

```diff
@@ -282,32 +281,30 @@ export class Project
       startupMessage: StartupMessage.Restarting,
     });
 
+    // we first consider forceCleanBuild flag, if set we always perform a clean
+    // start of the project and select the device
     if (forceCleanBuild) {
       await this.start(true, true);
       await this.selectDevice(deviceInfo, true);
       return;
     }
 
-    if (this.isCachedBuildStale) {
-      await this.selectDevice(deviceInfo, false);
-      return;
-    }
-
-    if (restartDevice) {
-      await this.selectDevice(deviceInfo, false);
-      return;
-    }
-
-    if (onlyReloadJSWhenPossible) {
-      // if reloading JS is possible, we try to do it first and exit in case of success
-      // otherwise we continue to restart using more invasive methods
-      if (await this.reloadMetro()) {
+    // Otherwise, depending on the project state we try deviceSelection, or
+    // only relad JS if possible.
+    try {
+      if (restartDevice || this.isCachedBuildStale) {
+        await this.selectDevice(deviceInfo, false);
         return;
       }
-    }
 
-    // otherwise we try to restart the device session
-    try {
+      if (onlyReloadJSWhenPossible) {
+        // if reloading JS is possible, we try to do it first and exit in case of success
+        // otherwise we continue to restart using more invasive methods
+        if (await this.reloadMetro()) {
```

Comment:
I don't like that logic, as it is right now unsuccessful metro reload causes full device reboot, which seems unnecessary if we know that the reason it is failing is because metro was terminated.

It seems to me that `onlyReloadJSWhenPossible` option should first cause `${metroURL}/reload` as it is right now, then if it fails it should run `this.metro.start()` and only if it fails it should reboot the device

Reply:
From my testing the metro process crashes are extremely rare, so it's not very often that it'll restart the device because of that.

Note that it isn't as simple to restart the metro w/o running a series of other reload tasks, because it launches on ephemeral port. So if we just restart it, the existing process wouldn't be able to connect to it anyway, as it'd bind on a different port.

Handling this properly would require some more in-depth refactor of device session and reset logic in general that I'd rather avoid, as we are aiming to stabilize the current codebase.

Reply:
I agree it would be to complex of a change for now